tmt/cli.py

```diff
@@ -1074,3 +1076,94 @@ def lint(context, **kwargs):
         except SystemExit as e:
             exit_code |= e.code
     raise SystemExit(exit_code)
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#  Setup
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+@main.group(cls=CustomGroup)
+@click.pass_context
+def setup(context, **kwargs):
+    """
+    Setup the environment for working with tmt.
+    """
+
+
+@setup.group(cls=CustomGroup)
+@click.pass_context
+def completion(context, **kwargs):
+    """
+    Setup shell completions.
+
+    By default, these commands only write a shell script to the output which
+    can then be sourced from the shell's configuration file.
+    """
+
+
+COMPLETE_VARIABLE = '_TMT_COMPLETE'
+COMPLETE_SCRIPT = 'tmt-complete'
+
+
+def setup_completion(shell, install):
+    """ Setup completion based on the shell """
+    config = tmt.utils.Config()
+    # Fish gets installed into its special location where it is automatically
+    # loaded.
+    if shell == 'fish':
+        script = os.path.expanduser('~/.config/fish/completions/tmt.fish')
+    # Bash and zsh get installed to tmt's config directory.
+    else:
+        script = os.path.join(config.path, f'{COMPLETE_SCRIPT}.{shell}')
+
+    out = open(script, 'w') if install else sys.stdout
+    subprocess.run(f'{COMPLETE_VARIABLE}={shell}_source tmt',
```

Comment:
Not completely sure if there is a better way to do this. I didn't find a way to instrument click to output the completion without ugly hacks/or rerunning tmt with the env variable set... Also wondering if it wouldn't be better to use `argv[0]` instead of `tmt`

Reply:
I think it's completely fine to re-run `tmt` with the environment variable to enable the `click` magic. Using `argv[0]` instead of hardcoded `tmt` might be a bit better.